plugins/mcreator-core/blockly/js/mcreator_extensions.js

```diff
@@ -198,4 +188,44 @@ function validateResourceLocationFields(...fields) {
 Blockly.Extensions.register('tag_input_field_validator', validateResourceLocationFields('tag'));
 
 Blockly.Extensions.register('geode_tag_fields_validator',
-    validateResourceLocationFields('cannot_replace_tag', 'invalid_blocks_tag'));
\ No newline at end of file
+    validateResourceLocationFields('cannot_replace_tag', 'invalid_blocks_tag'));
+
+Blockly.Extensions.registerMixin('controls_flow_in_loop_check_exclude_wait',
+    {
+        onchange: function (e) {
+            // Don't change state if it's at the start of a drag and it's not a move event
+            if (!this.workspace.isDragging || this.workspace.isDragging() || e.type !== Blockly.Events.BLOCK_MOVE) {
+                return;
+            }
+            const isWithinLoop = !!(checkIfWithin(this, function (type) {
+                return Blockly.libraryBlocks.loops.loopTypes.has(type);
+            }));
+            const isWithinWaitBlock = !!(checkIfWithin(this, function (type) {
+                return type === 'wait';
+            }));
+            if (!isWithinLoop) {
+                this.setWarningText(Blockly.Msg['CONTROLS_FLOW_STATEMENTS_WARNING']);
+            } else if (isWithinWaitBlock) {
+                this.setWarningText(javabridge.t('blockly.block.controls_flow_statements.inside_wait'));
+            } else {
+                this.setWarningText(null);
+            }
+            if (!this.isInFlyout) {
+                const group = Blockly.Events.getGroup();
+                // Makes it so the move and the disable event get undone together.
+                Blockly.Events.setGroup(e.group);
+                this.setEnabled(isWithinLoop && !isWithinWaitBlock);
+                Blockly.Events.setGroup(group);
+            }
+        }
+    });
+
+function checkIfWithin(block, predicate) {

```

Comment:
Rather than a predicate, I'd pass an array of string and check if the block type is in it (using `.includes(block.type)`)

Reply:
I wanted to make this more universal in case someone needs to check from something more specific than just a type